### ISSUE TITLE
Dataloader yield from root fiber with graphql-metrics

### ISSUE THREAD

**Issue by arathunku** [rmosolgo/graphql-ruby#3313]
**Describe the bug**

When graphql-metrics is used, dataloader yields from root fiber. This only happens for arguments when there are arguments with `loads: `

**Versions**

- `graphql` `1.12.3`
- `graphql-metrics` `4.0.6`

**GraphQL test script**

```ruby
require 'bundler/inline'

# $LOAD_PATH.unshift File.dirname(__FILE__) + '/lib'
# require_relative './lib/graphql'

gemfile do
  source 'https://rubygems.org'

  gem "graphql", "1.12.3"
  gem 'graphql-metrics', require: 'graphql/metrics'
  gem 'pry'
end

class BasicAnalyzer < GraphQL::Metrics::Analyzer
  def query_extracted(metrics)
  end
  def field_extracted(metrics)
  end
  def argument_extracted(metrics)
  end
  # Uncomment this to see that everything is OK
  # def on_leave_field(node, _parent, visitor)
  # end
end

class RecordLoader < GraphQL::Dataloader::Source
  def initialize(model)
    @model = model
  end

  def fetch(ids)
    self.class.build_things(ids)
  end

  def self.build_things(ids)
    ids.map { |id| OpenStruct.new(name: id, id: id) }
  end
end

class Thing < GraphQL::Schema::Object
  field :name, String, null: false

end

class Query < GraphQL::Schema::Object
  field :things, Thing.connection_type, null: false do
    argument(:parent_ids, [GraphQL::Types::ID], loads: Thing, required: true)
  end

  def things(parents:)
    context.dataloader.with(RecordLoader, ::Thing).load_all(['11'])
  end
end

class Schema < GraphQL::Schema
  query Query
  use GraphQL::Dataloader

  query_analyzer BasicAnalyzer

  def self.object_from_id(id, ctx)
    ctx.dataloader.with(RecordLoader, Thing).load(id)
  end

  def self.resolve_type(abst_type, obj, ctx)
    Thing
  end
end

pp Schema.execute('query { things(parentIds: ["11"]) { edges { node { name } } } }').to_h

end
```

**Steps to reproduce**

Run test script. Uncomment `#on_leave_field` in analyzer to see query run successfully.

**Expected behavior**

Dataloader not yielding in root fiber.

**Actual behavior**

What specifically went wrong?

Exception was raised.

Place full backtrace here (if a Ruby exception is involved):

<details>

<p>

```
Traceback (most recent call last):
        61: from test.rb:70:in `<main>'
        60: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema.rb:1661:in `execute'
        59: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema.rb:1690:in `multiplex'
        58: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/multiplex.rb:50:in `run_all'
        57: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/multiplex.rb:60:in `run_queries'
        56: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/tracing.rb:66:in `trace'
        55: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/multiplex.rb:62:in `block in run_queries'
        54: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/multiplex.rb:188:in `instrument_and_analyze'
        53: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:26:in `apply_instrumenters'
        52: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
        51: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:27:in `block in apply_instrumenters'
        50: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:44:in `each_query_call_hooks'
        49: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
        48: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:45:in `block in each_query_call_hooks'
        47: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:41:in `each_query_call_hooks'
        46: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:46:in `block (2 levels) in each_query_call_hooks'
        45: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/instrumentation.rb:29:in `block (2 levels) in apply_instrumenters'
        44: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/multiplex.rb:208:in `block in instrument_and_analyze'
        43: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast.rb:34:in `analyze_multiplex'
        42: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/tracing.rb:66:in `trace'
        41: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast.rb:35:in `block in analyze_multiplex'
        40: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast.rb:35:in `map'
        39: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast.rb:37:in `block (2 levels) in analyze_multiplex'
        38: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast.rb:61:in `analyze_query'
        37: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/tracing.rb:66:in `trace'
        36: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast.rb:73:in `block in analyze_query'
        35: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:41:in `visit'
        34: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:68:in `visit'
        33: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:174:in `on_node_with_modifications'
        32: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:79:in `visit_node'
        31: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:79:in `public_send'
        30: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:74:in `on_operation_definition'
        29: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:123:in `on_operation_definition'
        28: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:174:in `on_abstract_node'
        27: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:101:in `on_abstract_node'
        26: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:101:in `each'
        25: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:102:in `block in on_abstract_node'
        24: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:174:in `on_node_with_modifications'
        23: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:79:in `visit_node'
        22: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/language/visitor.rb:79:in `public_send'
        21: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:121:in `on_field'
        20: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:245:in `call_analyzers'
        19: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:245:in `each'
        18: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:247:in `block in call_analyzers'
        17: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/analysis/ast/visitor.rb:247:in `public_send'
        16: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-metrics-4.0.6/lib/graphql/metrics/analyzer.rb:42:in `on_leave_field'
        15: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/query.rb:256:in `arguments_for'
        14: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/interpreter/arguments_cache.rb:27:in `fetch'
        13: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/execution/interpreter/arguments_cache.rb:15:in `block (3 levels) in initialize'
        12: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:96:in `coerce_arguments'
        11: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:96:in `map'
        10: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:96:in `each'
         9: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:122:in `block in coerce_arguments'
         8: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:122:in `map'
         7: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:122:in `block (2 levels) in coerce_arguments'
         6: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:208:in `load_application_object'
         5: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/schema/member/has_arguments.rb:200:in `object_from_id'
         4: from test.rb:62:in `object_from_id'
         3: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/dataloader/source.rb:42:in `load'
         2: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/dataloader/source.rb:71:in `sync'
         1: from /home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/dataloader.rb:68:in `yield'
/home/arathunku/.asdf/installs/ruby/2.7.1/lib/ruby/gems/2.7.0/gems/graphql-1.12.3/lib/graphql/dataloader.rb:68:in `yield': can't yield from root fiber (FiberError)
```

</p>
</details>

<details>
<summary>Click to view exception backtrace</summary>

```
Something went wrong
2.6.0/gems/graphql-1.9.17/lib/graphql/subscriptions/instrumentation.rb:34:in `after_query'
… don't hesitate to include all the rows here: they will be collapsed
```

</details>

**Additional context**

I've tested this on master and on https://github.com/rmosolgo/graphql-ruby/pull/3312.

This problem didn't exist previously when GraphQL:::Batch was used. 

I'm unsure if this should be fixed in graphql-metrics or in graphql-ruby but given that this didn't exist before, I'm raising this here as a starting point. 

______

New Dataloader looks excellent!

**Comment by rmosolgo** [comment]
Thanks for the detailed bug report, I'll take a look soon! I bet this can (and should) be mitigated in GraphQL-Ruby.

**Comment by swalkinshaw** [comment]
cc @chrisbutcher in case it's on the metrics side of things

**Comment by rmosolgo** [comment]
No, I've got to handle this better in GraphQL-Ruby: analyzers can cause arguments to be `load: ...`ed, but they aren't wrapped in a Fiber by `GraphQL::Dataloader`. 

The good news is, #3312 makes dataloader much more general, and I bet that fixing _this_ bug will also fix the current dataloader caveat that arguments _aren't_ batch-loaded.

**Comment by rmosolgo** [comment]
:+1: Fixed on #3320: 

```
% ruby analyzer_test.rb
{"data"=>{"things"=>{"edges"=>[{"node"=>{"name"=>"11"}}]}}}
```

I still have some work to do on that PR, but getting there!

**Comment by arathunku** [comment]
This indeed solves the issue. It also looks like https://github.com/rmosolgo/graphql-ruby/issues/3314 is fixed too and all our internal specs pass. Thank you @rmosolgo!!!
